appinventor/blocklyeditor/src/generators/yail/text.js

```diff
@@ -318,3 +318,30 @@ Blockly.Yail['text_is_string'] = function() {
   code = code + Blockly.Yail.YAIL_DOUBLE_QUOTE + "is a string?" + Blockly.Yail.YAIL_DOUBLE_QUOTE + Blockly.Yail.YAIL_CLOSE_COMBINATION;
   return [ code, Blockly.Yail.ORDER_ATOMIC ];
 };
+
+Blockly.Yail['text_replace_mappings'] = function() {
+    // String replace with segment
+    var argument0 = Blockly.Yail.valueToCode(this, 'TEXT', Blockly.Yail.ORDER_NONE) || "\"\"";
+    var argument1 = Blockly.Yail.valueToCode(this, 'MAPPINGS', Blockly.Yail.ORDER_NONE) || "\"\"";
+    var mode = this.getFieldValue('OP');
+    var argument2 = Blockly.Yail.text_replace_mappings.OPERATORS[mode];
+
+    var code = Blockly.Yail.YAIL_CALL_YAIL_PRIMITIVE + "string-replace-mappings" + Blockly.Yail.YAIL_SPACER;
+    code = code + Blockly.Yail.YAIL_OPEN_COMBINATION + Blockly.Yail.YAIL_LIST_CONSTRUCTOR + Blockly.Yail.YAIL_SPACER;
+    code = code + argument0 + Blockly.Yail.YAIL_SPACER;
+    code = code + argument1 + Blockly.Yail.YAIL_SPACER;
+    code = code + argument2 + Blockly.Yail.YAIL_CLOSE_COMBINATION;
+    code = code + Blockly.Yail.YAIL_SPACER + Blockly.Yail.YAIL_QUOTE;
+    code = code + Blockly.Yail.YAIL_OPEN_COMBINATION + "text" + Blockly.Yail.YAIL_SPACER;
+    code = code + "dictionary" + Blockly.Yail.YAIL_SPACER;
+    code = code + "number" + Blockly.Yail.YAIL_CLOSE_COMBINATION;
+    code = code + Blockly.Yail.YAIL_SPACER + Blockly.Yail.YAIL_DOUBLE_QUOTE + "replace with mappings" + Blockly.Yail.YAIL_DOUBLE_QUOTE + Blockly.Yail.YAIL_CLOSE_COMBINATION;
+
+    return [ code, Blockly.Yail.ORDER_ATOMIC ];
+  };
+
+  Blockly.Yail.text_replace_mappings.OPERATORS = {
```

Comment:
@ewpatton Right now what I did was I took the order as a String parameter, and mapped it to an integer, which I pass in to the corresponding Java function as a parameter later on. Would this be a viable approach, or should we define a function for each mode in Scheme, and then call in to the appropriate function?

Reply:
I think it ultimately depends on how much code overlap there is. If it's saner to keep each as a different function you could map to three Scheme functions that call through to the Java implementation. If it's easier to pass an argument, then I would say pass the String itself and not bother mapping to an integer internally.

Reply:
@ewpatton I figure keeping separate functions is a bit cleaner (since we do some form of mapping in the generator anyway). I did some refactoring; Let me know if you think it's overkill (I introduce a class for ordering in JavaStringUtils)